outlines/base.py

```diff
@@ -0,0 +1,204 @@
+    chararrays and returns a single NumPy chararrays or a tuple of NumPy chararrays.
+
+    Its behavior is similar to NumPy's `vectorize` function for Python functions,
+    and uses `async.gather` for faster execution of coroutines.
+
+    Part of the code was reproduced from `numpy.lib.function_base`.
+
+    """
+
+    def __init__(self, func: Callable, signature: Optional[str] = None):
+        self.func = func
+        self.signature = signature
+        self.is_coroutine_fn = inspect.iscoroutinefunction(func)
+
+        self.__name__ = func.__name__
+        if hasattr(func, "__doc__"):
+            self.__doc__ = func.__doc__
+
+        if signature is not None:
+            self._in_and_out_core_dimensions = _parse_gufunc_signature(signature)
+        else:
+            self._in_and_out_core_dimensions = None
+
+    def __call__(self, *args, **kwargs):
+        if self.signature is not None:
+            return self.call_with_signature(*args, **kwargs)
+        elif not args and not kwargs:  # Execute thunks directly
+            if self.is_coroutine_fn:
+                loop = asyncio.new_event_loop()
+                try:
+                    return loop.run_until_complete(self.func())
+                finally:
+                    loop.close()
+            else:
+                return self.func()
+        else:
+            return self.call_no_signature(*args, **kwargs)
+
+    def call_no_signature(self, *args, **kwargs):
+        # Convert args and kwargs to arrays
+        args = [np.array(arg) for arg in args]
+        kwargs = {key: np.array(value) for key, value in kwargs.items()}
+
+        # Broadcast args and kwargs
+        broadcast_shape = np.broadcast(*args, *list(kwargs.values())).shape
+        args = [np.broadcast_to(arg, broadcast_shape) for arg in args]
+        kwargs = {
+            key: np.broadcast_to(value, broadcast_shape)
+            for key, value in kwargs.items()
+        }
+
+        # Flatten args and kwargs
+        flat_args = [np.ravel(arg) for arg in args]
+        flat_kwargs = {key: np.ravel(value) for key, value in kwargs.items()}
+
+        if self.is_coroutine_fn:
+            loop = asyncio.new_event_loop()
+            try:
+                outputs = loop.run_until_complete(
+                    self.vectorize_call_coroutine(
+                        broadcast_shape, flat_args, flat_kwargs
+                    )
+                )
+            finally:
+                loop.close()
+        else:
+            outputs = self.vectorize_call(broadcast_shape, flat_args, flat_kwargs)
+
+        if isinstance(outputs[0], tuple):
+            outputs = tuple(
+                [
+                    np.asarray(x).reshape(broadcast_shape).squeeze()
+                    for x in zip(*outputs)
+                ]
+            )
+        else:
+            outputs = np.asarray(outputs).reshape(broadcast_shape).squeeze()
+
+        return outputs
+
+    def call_with_signature(self, *args, **kwargs):
```

Comment:
We should document this function for future contributors as this is not completely trivial